### ISSUE TITLE
[Backport v2.6] [BUG] Deleted cluster not removed from UI view until hard refresh (Ghost cluster)

### ISSUE THREAD

**Issue by janeczku** [rancher/rancher#36072]
Rancher 2.6.1

1. Login as standard, non-admin user
2. Create a RKE2 cluster
3. In a second browser login as admin user and navigate to Cluster Manager view
4. As the non-admin user delete the cluster 

**Expected:**

When the cluster is deleted it is removed from the view.

**Actual:**

Normal user session: Deleted cluster is not removed from the view until performing a hard browser refresh.
Admin user session:  Deleted cluster is removed from the view immediately

https://user-images.githubusercontent.com/3813921/137932366-12f9f499-0e30-4c17-86b5-b361ba95c5a5.mov

````
2021/10/19 15:03:11 [INFO] rkecluster fleet-default/template-rke2: waiting for at least one bootstrap node
2021/10/19 15:03:11 [INFO] [mgmt-auth-crtb-controller] Deleting roleBinding crb-n4zbcfymqv
2021/10/19 15:03:11 [INFO] [mgmt-auth-crtb-controller] Deleting rolebinding creator-cluster-owner-cluster-owner in namespace p-l9xs6 for crtb creator-cluster-owner
2021/10/19 15:03:11 [INFO] [mgmt-auth-crtb-controller] Deleting rolebinding creator-cluster-owner-cluster-owner in namespace p-hhc6q for crtb creator-cluster-owner
2021/10/19 15:03:12 [INFO] [mgmt-project-rbac-remove] Deleting namespace p-l9xs6
2021/10/19 15:03:12 [INFO] [mgmt-project-rbac-remove] Deleting namespace p-hhc6q
2021/10/19 15:03:16 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:16 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:16 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:16 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:16 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:16 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:17 [ERROR] error syncing 'fleet-default/template-rke2-dsfsdfsdfsdf-7ccf7c9f-98fgb': handler machine-provision-remove: cannot delete machine template-rke2-dsfsdfsdfsdf-7ccf7c9f-98fgb because create job has not finished, requeuing
2021/10/19 15:03:17 [INFO] rkecluster fleet-default/template-rke2: waiting for at least one bootstrap node
2021/10/19 15:03:17 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:17 [INFO] [mgmt-auth-prtb-controller] Updating owner label for roleBinding crb-mrk3zsur4y
2021/10/19 15:03:17 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:18 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: clusterregistrationtokens.management.cattle.io "default-token" is forbidden: unable to create new content in namespace c-m-v8p56lnd because it is being terminated, requeuing
2021/10/19 15:03:18 [INFO] [mgmt-auth-prtb-controller] Deleting roleBinding crb-mrk3zsur4y
2021/10/19 15:03:19 [INFO] [mgmt-cluster-rbac-delete] Creating namespace c-m-v8p56lnd
2021/10/19 15:03:19 [ERROR] error syncing 'c-m-v8p56lnd': handler cluster-watch: namespaces "c-m-v8p56lnd" not found, requeuing
2021/10/19 15:03:20 [ERROR] error syncing 'harvey': handler machine-worker-label: machines.cluster.x-k8s.io "custom-599f50dd418f" not found, requeuing
```

**Comment by gaktive** [comment]
@Jono-SUSE-Rancher fyi

**Comment by MbolotSuse** [comment]
## Root Cause

When a cluster was deleted, the roles/role-bindings which gave non-admin users permission to view the state of the cluster was immediately deleted. Since these users no longer had access to the state of the cluster, they were not able to view the changing status of the cluster. This prevented the UI from realizing that the cluster was deleted, and therefore should be removed from view for the user.

## What was fixed, or what change have occurred
Roles/Rolebindings in the local cluster now have an `OnRemove` method which checks the state of the cluster that they are related to on delete. If the cluster was not found, the function returns without error, allowing the role/binding to be deleted. However, if the role was found, the controller re-enqueues with an error (errSkip which should not leave any log messages) requiring the function to be run again until the function exists successfully.

Until the function exists successfully, a finalizer is left on the object (a byproduct of the controller) which prevents the role/binding from being deleted.

## Areas or cases that should be tested

1. The original scenario (repeated below).
2. Cluster deletes - we should ensure that roles/bindings related to the cluster are removed and that no "orphan bindings" are left over.
3. Role assignment deletes (cluster/project). If a user is unbound from a role, rolebindings which granted permissions in the cluster should also be removed and the user should no longer have those permissions (i.e. Create a cluster -> Assign a user as a cluster member -> Remove user as a cluster member -> ensure permissions are lost and bindings are deleted).
4. Role Template Deletes (cluster/project). If a roletemplate is deleted, all roles created for that roletemplate (and bindings to those roles) should be deleted (i.e. Create a new cluster/project role -> assign a user to that role -> delete cluster/project role -> ensure permissions are lost and bindings are deleted).

## What areas could experience regressions

Role binding end of lifecycle (delete), particularly surrounding cluster deletion, template deletion, and binding removal (see above).

## Are the repro steps accurate/minimal?
Yes. They are included here for convenience.
1. Run rancher/rancher:v2.6.3 (docker or HA).
2. Create a downstream cluster (tested/validated with DigitalOcean RKE1 - though this should not matter).
3. Create a new local user with a global role of `Standard User`.
4. Add the new local user to the downstream cluster as a `Cluster Member`.
5. In a new private window, login as the new user. Validate that you can see the cluster in the `Cluster Management` page accessible from the hamburger menu.
6. With the private window and the admin window side-by-side, delete the downstream cluster. Observer that the cluster disappeared from the admin user's page, but remained on the standard user's page indefinitely, requiring a refresh (see video in the issue).

**Comment by anupama2501** [comment]
Unable to list clusters as a standard users https://github.com/rancher/dashboard/issues/5711

**Comment by anupama2501** [comment]
Re-opening 

Validated the following scenarios on v2.6.3 upgraded to v2.6-head 312d8bb

1. Created 2 RKE2 node driver and 2 RKE1 node driver clusters on 2.6.3 - rke2-1, rke2-2, rke1-3, rke1-4
3. Create 3 users - u1, u2 ,u3. Add u1 as Cluster owner, u2 as cluster member and u3 with custom permissions - `view nodes`
4. Delete rke2-1 and rke1-3 clusters and noticed the cluster is stuck on removing as cluster member 
5. Upgrade rancher to v2.6-head
6. As a cluster owner delete the cluster rke2-2 and rke1-4.
7. Still see the rke2-2 cluster is stuck in removing for a cluster member and is only removed upon refresh.
<img width="500" alt="deleterke2Removing" src="https://user-images.githubusercontent.com/38144301/165886313-21e1046a-e133-47a2-baba-182ae1a3509a.png">

**Comment by anupama2501** [comment]
Issue is resolved for RKE1 clusters as they not stuck in removing for cluster members and are deleted at the same time as of admin.

**Comment by steffantucker** [comment]
After some investigation, this appears to be a frontend issue. The issue seems specific to cluster owners though, but I have figured out you can resurrect the removed cluster for at least an admin.
As above, I:
1. Create a standard user 
 - create another user to add as a member
3. As the standard user, create a new cluster
 - add the other user as a member of the cluster
4. Delete the cluster
For any/all users that are not cluster owners, the cluster will show removing, then disappear from the dashboard home page and the cluster management page.
In addition to this, I've found that at least for an admin user, if they haven't refreshed the page, if another cluster is created that had the same name as the deleted cluster, that deleted cluster will show up again but only on the home page, and will be in the same "Removing" state experienced by cluster owners until the page is refreshed. I've also verified it's affecting all cluster owners by creating additional users and adding them as cluster owners, all cluster owners get stuck in the same state.
What has led me to this conclusion is that the page does receive a request telling it to remove the cluster, follows is the JSON received by the front end:
```json
{
    "name": "resource.remove",
    "revision": "2604427",
    "data": {
        "id": "c-m-c22gtx49",
        "type": "management.cattle.io.cluster",
        "links": {
            "log": "https://rancher.147.182.209.61.sslip.io/v1/management.cattle.io.clusters/c-m-c22gtx49?link=log",
            "projects": "https://rancher.147.182.209.61.sslip.io/v1/management.cattle.io.clusters/c-m-c22gtx49?link=projects",
            "remove": "blocked",
            "schemas": "https://rancher.147.182.209.61.sslip.io/v1/management.cattle.io.clusters/c-m-c22gtx49?link=schemas",
            "self": "https://rancher.147.182.209.61.sslip.io/v1/management.cattle.io.clusters/c-m-c22gtx49",
            "shell": "wss://rancher.147.182.209.61.sslip.io/v3/clusters/c-m-c22gtx49?shell=true",
            "subscribe": "https://rancher.147.182.209.61.sslip.io/v1/management.cattle.io.clusters/c-m-c22gtx49?link=subscribe",
            "update": "blocked",
            "view": "https://rancher.147.182.209.61.sslip.io/apis/management.cattle.io/v3/clusters/c-m-c22gtx49"
        },
        "actions": {
            "apply": "https://rancher.147.182.209.61.sslip.io/v1/management.cattle.io.clusters/c-m-c22gtx49?action=apply",
            "enableMonitoring": "https://rancher.147.182.209.61.sslip.io/v3/clusters/c-m-c22gtx49?action=enableMonitoring",
            "generateKubeconfig": "https://rancher.147.182.209.61.sslip.io/v3/clusters/c-m-c22gtx49?action=generateKubeconfig",
            "importYaml": "https://rancher.147.182.209.61.sslip.io/v3/clusters/c-m-c22gtx49?action=importYaml"
        },
        "apiVersion": "management.cattle.io/v3",
        "kind": "Cluster",
        "metadata": {
            "annotations": {
                "authz.management.cattle.io/creator-role-bindings": "{\"created\":[\"cluster-owner\"],\"required\":[\"cluster-owner\"]}",
                "field.cattle.io/creatorId": "u-vdlww",
                "lifecycle.cattle.io/create.cluster-agent-controller-cleanup": "true",
                "lifecycle.cattle.io/create.cluster-provisioner-controller": "true",
                "lifecycle.cattle.io/create.cluster-scoped-gc": "true",
                "lifecycle.cattle.io/create.mgmt-cluster-rbac-remove": "true",
                "objectset.rio.cattle.io/applied": "H4sIAAAAAAAA/4xSQW/bPAz9Kx94tt0vSZslBnYoih2KYV0Pw3ZmLNrmIlMGRdsrivz3QW7SZh1a7CRI4nt8j4+PgD1/J40cBEroULChjsSKCs08FRwuxhVksGdxUMKNH6KRQgYdGTo0hPIRUCQYGgeJ6VozeXdGUCmhBb1NBEM+Oj9NkEHY/aTKIlmhHM6qOZVVT33yGUrvVIdJSPNm3EMJvYaRkxOW5lz/IvvvM4v7+CL+fTbBjl40/Hn8Ezj2WCWG2hNZ7qjGwRtkbwlE17FwNEWjZN50IDhkMJvnIN+4o2jY9VDK4H0GHnfk51G/paXF2EIJS7febNd06Rx9qFe0XmycWy2u3LZ2iPXVrl5s3WZVL1O3k+u8y6vlsrFfl9v0HHuq5owbErvtsKGvI6myS8WQAUqcSJOYQwaOYqXc29M2wfzASu76GQwlKErVkl4cz3xmLsdlsS7WZ5DB2teI/bCjHHvOcbC2HP8vFsUmITj2Hh/u3o+NBHeejjtw7UmNpYGyRh/p1e+XIGxB//6/I5uC7u+D5+rhlMac8o+g+zn2o4rX0XPXBzVyN0FqbtJAk5nTDdKkWYxU0D/39KFCfxI8WPtJXB9YLKGf9Lhj7SGDicWFKd4r1aTkTrt+/D/8DgAA///6gRUw6QMAAA",
                "objectset.rio.cattle.io/id": "cluster-create",
                "objectset.rio.cattle.io/owner-gvk": "provisioning.cattle.io/v1, Kind=Cluster",
                "objectset.rio.cattle.io/owner-name": "clusterclustercluster",
                "objectset.rio.cattle.io/owner-namespace": "fleet-default",
                "provisioning.cattle.io/administrated": "true"
            },
            "creationTimestamp": "2022-07-26T21:49:40Z",
            "deletionGracePeriodSeconds": 0,
            "deletionTimestamp": "2022-07-26T21:59:53Z",
            "fields": [
                "c-m-c22gtx49",
                "10m"
            ],
            "finalizers": [
                "controller.cattle.io/mgmt-cluster-rbac-remove"
            ],
            "generation": 35,
            "labels": {
                "objectset.rio.cattle.io/hash": "2d6896e4dde7f3e618dd315d9fdaaf5bf19d83f2",
                "provider.cattle.io": "rke2"
            },
            "managedFields": [
                {
                    "apiVersion": "management.cattle.io/v3",
                    "fieldsType": "FieldsV1",
                    "fieldsV1": {
                        "f:metadata": {
                            "f:annotations": {
                                ".": {},
                                "f:authz.management.cattle.io/creator-role-bindings": {},
                                "f:field.cattle.io/creatorId": {},
                                "f:lifecycle.cattle.io/create.cluster-agent-controller-cleanup": {},
                                "f:lifecycle.cattle.io/create.cluster-provisioner-controller": {},
                                "f:lifecycle.cattle.io/create.cluster-scoped-gc": {},
                                "f:lifecycle.cattle.io/create.mgmt-cluster-rbac-remove": {},
                                "f:objectset.rio.cattle.io/applied": {},
                                "f:objectset.rio.cattle.io/id": {},
                                "f:objectset.rio.cattle.io/owner-gvk": {},
                                "f:objectset.rio.cattle.io/owner-name": {},
                                "f:objectset.rio.cattle.io/owner-namespace": {},
                                "f:provisioning.cattle.io/administrated": {}
                            },
                            "f:finalizers": {
                                ".": {},
                                "v:\"controller.cattle.io/mgmt-cluster-rbac-remove\"": {}
                            },
                            "f:labels": {
                                ".": {},
                                "f:objectset.rio.cattle.io/hash": {},
                                "f:provider.cattle.io": {}
                            }
                        },
                        "f:spec": {
                            ".": {},
                            "f:agentImageOverride": {},
                            "f:answers": {},
                            "f:description": {},
                            "f:desiredAgentImage": {},
                            "f:desiredAuthImage": {},
                            "f:displayName": {},
                            "f:enableClusterAlerting": {},
                            "f:enableClusterMonitoring": {},
                            "f:enableNetworkPolicy": {},
                            "f:fleetWorkspaceName": {},
                            "f:importedConfig": {
                                ".": {},
                                "f:kubeConfig": {}
                            },
                            "f:internal": {},
                            "f:localClusterAuthEndpoint": {
                                ".": {},
                                "f:enabled": {}
                            },
                            "f:windowsPreferedCluster": {}
                        },
                        "f:status": {
                            ".": {},
                            "f:agentFeatures": {
                                ".": {},
                                "f:embedded-cluster-api": {},
                                "f:fleet": {},
                                "f:monitoringv1": {},
                                "f:multi-cluster-management": {},
                                "f:multi-cluster-management-agent": {},
                                "f:provisioningv2": {},
                                "f:rke2": {}
                            },
                            "f:agentImage": {},
                            "f:aksStatus": {
                                ".": {},
                                "f:privateRequiresTunnel": {},
                                "f:rbacEnabled": {},
                                "f:upstreamSpec": {}
                            },
                            "f:allocatable": {
                                ".": {},
                                "f:cpu": {},
                                "f:memory": {},
                                "f:pods": {}
                            },
                            "f:apiEndpoint": {},
                            "f:appliedAgentEnvVars": {},
                            "f:appliedEnableNetworkPolicy": {},
                            "f:appliedPodSecurityPolicyTemplateId": {},
                            "f:appliedSpec": {
                                ".": {},
                                "f:agentImageOverride": {},
                                "f:answers": {},
                                "f:description": {},
                                "f:desiredAgentImage": {},
                                "f:desiredAuthImage": {},
                                "f:displayName": {},
                                "f:enableClusterAlerting": {},
                                "f:enableClusterMonitoring": {},
                                "f:enableNetworkPolicy": {},
                                "f:internal": {},
                                "f:localClusterAuthEndpoint": {
                                    ".": {},
                                    "f:enabled": {}
                                },
                                "f:windowsPreferedCluster": {}
                            },
                            "f:authImage": {},
                            "f:caCert": {},
                            "f:capabilities": {
                                ".": {},
                                "f:loadBalancerCapabilities": {}
                            },
                            "f:capacity": {
                                ".": {},
                                "f:cpu": {},
                                "f:memory": {},
                                "f:pods": {}
                            },
                            "f:conditions": {},
                            "f:driver": {},
                            "f:eksStatus": {
                                ".": {},
                                "f:managedLaunchTemplateID": {},
                                "f:managedLaunchTemplateVersions": {},
                                "f:privateRequiresTunnel": {},
                                "f:securityGroups": {},
                                "f:subnets": {},
                                "f:upstreamSpec": {},
                                "f:virtualNetwork": {}
                            },
                            "f:gkeStatus": {
                                ".": {},
                                "f:privateRequiresTunnel": {},
                                "f:upstreamSpec": {}
                            },
                            "f:limits": {
                                ".": {},
                                "f:cpu": {},
                                "f:memory": {},
                                "f:pods": {}
                            },
                            "f:linuxWorkerCount": {},
                            "f:nodeCount": {},
                            "f:provider": {},
                            "f:requested": {
                                ".": {},
                                "f:cpu": {},
                                "f:memory": {},
                                "f:pods": {}
                            },
                            "f:serviceAccountToken": {},
                            "f:version": {
                                ".": {},
                                "f:buildDate": {},
                                "f:compiler": {},
                                "f:gitCommit": {},
                                "f:gitTreeState": {},
                                "f:gitVersion": {},
                                "f:goVersion": {},
                                "f:major": {},
                                "f:minor": {},
                                "f:platform": {}
                            }
                        }
                    },
                    "manager": "rancher",
                    "operation": "Update",
                    "time": "2022-07-26T21:57:32Z"
                }
            ],
            "name": "c-m-c22gtx49",
            "relationships": [
                {
                    "fromId": "fleet-default/clusterclustercluster",
                    "fromType": "provisioning.cattle.io.cluster",
                    "rel": "applies"
                },
                {
                    "toId": "c-m-c22gtx49-clusterowner",
                    "toType": "rbac.authorization.k8s.io.clusterrole",
                    "rel": "owner",
                    "state": "active",
                    "message": "Resource is current"
                },
                {
                    "toId": "c-m-c22gtx49-clustermember",
                    "toType": "rbac.authorization.k8s.io.clusterrole",
                    "rel": "owner",
                    "state": "active",
                    "message": "Resource is current"
                },
                {
                    "toId": "c-m-c22gtx49-restricted-admin-cr-clusters",
                    "toType": "rbac.authorization.k8s.io.clusterrole",
                    "rel": "owner",
                    "state": "active",
                    "message": "Resource is current"
                }
            ],
            "resourceVersion": "2604427",
            "state": {
                "error": false,
                "message": "Resource scheduled for deletion",
                "name": "removing",
                "transitioning": true
            },
            "uid": "e3b9f2bd-9d82-4f84-a184-e57134b4cab0"
        },
        "spec": {
            "agentImageOverride": "",
            "answers": {},
            "description": "",
            "desiredAgentImage": "rancher/rancher-agent:v2.6.6",
            "desiredAuthImage": "rancher/kube-api-auth:v0.1.8",
            "displayName": "clusterclustercluster",
            "enableClusterAlerting": false,
            "enableClusterMonitoring": false,
            "enableNetworkPolicy": null,
            "fleetWorkspaceName": "fleet-default",
            "importedConfig": {
                "kubeConfig": ""
            },
            "internal": false,
            "localClusterAuthEndpoint": {
                "enabled": false
            },
            "windowsPreferedCluster": false
        },
        "status": {
            "agentFeatures": {
                "embedded-cluster-api": false,
                "fleet": false,
                "monitoringv1": false,
                "multi-cluster-management": false,
                "multi-cluster-management-agent": true,
                "provisioningv2": false,
                "rke2": false
            },
            "agentImage": "rancher/rancher-agent:v2.6.6",
            "aksStatus": {
                "privateRequiresTunnel": null,
                "rbacEnabled": null,
                "upstreamSpec": null
            },
            "allocatable": {
                "cpu": "2",
                "memory": "4026032Ki",
                "pods": "110"
            },
            "apiEndpoint": "https://10.43.0.1:443",
            "appliedAgentEnvVars": [
                {
                    "name": "CATTLE_SERVER_VERSION",
                    "value": "v2.6.6"
                },
                {
                    "name": "CATTLE_INSTALL_UUID",
                    "value": "3d884ed9-3f1f-4a51-ade2-ce0bbc60e2f0"
                },
                {
                    "name": "CATTLE_INGRESS_IP_DOMAIN",
                    "value": "sslip.io"
                }
            ],
            "appliedEnableNetworkPolicy": false,
            "appliedPodSecurityPolicyTemplateId": "",
            "appliedSpec": {
                "agentImageOverride": "",
                "answers": {},
                "description": "",
                "desiredAgentImage": "",
                "desiredAuthImage": "",
                "displayName": "",
                "enableClusterAlerting": false,
                "enableClusterMonitoring": false,
                "enableNetworkPolicy": null,
                "internal": false,
                "localClusterAuthEndpoint": {
                    "enabled": false
                },
                "windowsPreferedCluster": false
            },
            "authImage": "",
            "caCert": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJlRENDQVIrZ0F3SUJBZ0lCQURBS0JnZ3Foa2pPUFFRREFqQWtNU0l3SUFZRFZRUUREQmx5YTJVeUxYTmwKY25abGNpMWpZVUF4TmpVNE9EY3lNamszTUI0WERUSXlNRGN5TmpJeE5URXpOMW9YRFRNeU1EY3lNekl4TlRFegpOMW93SkRFaU1DQUdBMVVFQXd3WmNtdGxNaTF6WlhKMlpYSXRZMkZBTVRZMU9EZzNNakk1TnpCWk1CTUdCeXFHClNNNDlBZ0VHQ0NxR1NNNDlBd0VIQTBJQUJPNEI3OE5PTUFiZldQdDJoQnlJM242bmJPMXlhU1NDVlY4cFBSelAKWlZwMEsxdGtIR0RlMjUwS2tQU2NsQjRmSkR1UnE1VUE5MlN3L2p2Mnl5dE1Ha3lqUWpCQU1BNEdBMVVkRHdFQgovd1FFQXdJQ3BEQVBCZ05WSFJNQkFmOEVCVEFEQVFIL01CMEdBMVVkRGdRV0JCUk5zUlAyYjhtNWNxNFc2cGs4ClJBRDZBZ3NJcWpBS0JnZ3Foa2pPUFFRREFnTkhBREJFQWlCRFdIa2lGWlQwYWN1UXhZbDYwU2JmS1I3bkNSdTYKQ0QrZlpuWEMrQnp2Z0FJZ1o5SEpRSDd0UWQwckNNKytCWHNnd2tkZlJEU1BpQmN6ZDJaRXRXNlNEUWs9Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K",
            "capabilities": {
                "loadBalancerCapabilities": {}
            },
            "capacity": {
                "cpu": "2",
                "memory": "4026032Ki",
                "pods": "110"
            },
            "conditions": [
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:40Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "BackingNamespaceCreated"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:41Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "DefaultProjectCreated"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:41Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "SystemProjectCreated"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:41Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "InitialRolesPopulated"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:56:56Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "Connected"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:44Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "CreatorMadeOwner"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:50:52Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "Provisioned"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:53Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "NoDiskPressure"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:53Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "NoMemoryPressure"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:49:54Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "SecretsMigrated"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:57:31Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "Updated"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:57:03Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "Ready"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:56:49Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "GlobalAdminsSynced"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:57:03Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "Waiting"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:57:13Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "SystemAccountCreated"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T21:57:18Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "AgentDeployed"
                },
                {
                    "error": false,
                    "lastUpdateTime": "2022-07-26T22:00:04Z",
                    "status": "True",
                    "transitioning": false,
                    "type": "Removed"
                }
            ],
            "driver": "imported",
            "eksStatus": {
                "managedLaunchTemplateID": "",
                "managedLaunchTemplateVersions": null,
                "privateRequiresTunnel": null,
                "securityGroups": null,
                "subnets": null,
                "upstreamSpec": null,
                "virtualNetwork": ""
            },
            "gkeStatus": {
                "privateRequiresTunnel": null,
                "upstreamSpec": null
            },
            "limits": {
                "cpu": "120m",
                "memory": "148Mi",
                "pods": "0"
            },
            "linuxWorkerCount": 1,
            "nodeCount": 1,
            "provider": "rke2",
            "requested": {
                "cpu": "1220m",
                "memory": "238Mi",
                "pods": "17"
            },
            "serviceAccountToken": <removed>,
            "version": {
                "buildDate": "2022-06-08T18:03:44Z",
                "compiler": "gc",
                "gitCommit": "42c05a547468804b2053ecf60a3bd15560362fc2",
                "gitTreeState": "clean",
                "gitVersion": "v1.23.7+rke2r2",
                "goVersion": "go1.17.10b7",
                "major": "1",
                "minor": "23",
                "platform": "linux/amd64"
            }
        }
    }
}
```

**Comment by MbolotSuse** [comment]
@steffantucker I don't believe this to be a frontend issue. Our current model of receiving updates uses websockets - essentially the backend sends the user an update with new information about the cluster. When looking at this before, the conclusion was that the user looses permissions to see the state of the cluster before it is removed. The UI then keeps the cluster up because it has not received any information indicating that the cluster was deleted.

To explain in full, here's the process for each user:

Standard User:
- The user can see the management.cattle.io cluster object because they can get that specific cluster
- The cluster is deleted
- The user's permissions to view that specific cluster are removed
- The user does not receive a delete event
- The UI is not cleared

Admin:
- The user can see the management.cattle.io cluster object because the admin can see * on * (everything)
- The cluster is deleted
- The admin's permissions remain unchanged
- Since the admin has * on *, they receive the delete event
- The UI sees the delete event and removes the cluster

Feel free to reach out with any questions.

**Comment by sowmyav27** [comment]
@steffantucker Can you please provide more info in the QA Test template? especially anything around `QA Testing Considerations` and `Regression areas` ?

**Comment by MbolotSuse** [comment]
## Validation Template

This template is for the second version of the fix. Information here supersedes what was in the the previous validation template. 

## Root Cause

Users are assigned to particular roles/roleBindings (for provisioning clusters) and clusterRoles/clusterRoleBindings(for management clusters) which give them permission to view the cluster object in the local cluster. These roles are owned by the cluster object (in part) and are deleted when the cluster is deleted. In the current version of rancher, this results in these objects being removed before the cluster is completely gone. This causes users to not receive (through the websocket) the delete events for the cluster, which results in the cluster remaining on the UI. Admin users, which have `*` verbs on `*` resources in `*` groups in the local cluster, do not suffer from these issues, resulting in the discrepancy pointed out in the issue.

The root cause of the previous fix being insufficient for RKE2 clusters was that v2 provisioning clusters use a different CRD as their primary cluster object (`clusters.provisioning.cattle.io`). These cluster types were not checked and the roles which gave access to these still experienced the issues outlined above.

## What was fixed, or what change have occurred

Logic has been added which blocks roles, role bindings, cluster roles, and cluster rolebindings which grant access to a cluster from being deleted as long as the cluster is still in a deleting state (i.e. can be retrieved from k8s and has a non-nil deletion timestamp).

The previous version of this change only considered mgmt type clusters (`clusters.management.cattle.io`). This new change considers provisioning type clusters (`clusters.provisioning.cattle.io`) and blocks deletion of the RBAC objects outlined above so long as clusters of either type are in a deleting state.

## Areas or cases that should be tested

### Notes 
- The functionality below relies on the websocket remaining active throughout the cluster deletion process. If the websocket is interrupted during deletion, the cluster will remain on the UI. This issue is not related to our RBAC or this fix, and any attempt to fix our websocket stability will need to occur during a separate PR.
- The changes made in this PR affect the ability of roles, cluster roles, role bindings, and cluster role bindings, to delete. In every test case, it should be validated that the RBAC objects were able to be deleted and that we didn't end up creating new "orphaned" objects. See the section on regressions for more details.

### Test Cases
1. The original issue with RKE1 and RKE2 clusters. These steps should be re-run with a new user that has a variety of permissions in the downstream cluster (both cluster level like `cluster-owner`, `cluster-member`, `view nodes`, etc. and project level like `project owner`, `manage project members`, etc.). General Steps:
- Run rancher
- Create a new user with `standard user` global role
- Create a new cluster, as admin
- Add the user to the cluster as a cluster owner
- Delete the cluster. Observe the UI to ensure that the cluster is removed from the window of the standard user and the admin user.
- Verify that no RBAC resources exist which give permission to view this cluster in the local cluster (i.e. no roles/clusterroles which grant any verbs on specifically this cluster, and no rolebindings/clusterrolebindings to such roles/cluster roles or to roles/cluster roles which would have given such access but now no longer exist)

2. Removal of a user from a cluster on RKE1/RKE2 clusters. Steps:
- Run rancher
- Create a new user with `standard user` global role
- Create a new cluster, as admin
- Add the user to the cluster as a cluster owner
- Remove the user from the cluster owner role
- Validate that the user cannot see the cluster any longer
- Validate that no RBAC resources exist in the local cluster which give users access to the management or provisioning cluster objects for this cluster.

3. Removal of a user from a project on RKE1/RKE2 clusters. Steps:
- Run rancher
- Create a new user with `standard user` global role
- Create a new cluster, as admin
- In the cluster, create a new project
- Add the new user as a project owner to this cluster
- Validate that the new user can see the cluster in the UI
- Remove the user from the project owner role
- Validate that the user cannot see the cluster any longer
- Validate that no RBAC resources exist in the local cluster which give users access to the management or provisioning cluster objects for this cluster.

4. Local cluster basic RBAC test. Steps:
- Run rancher. 
- In the local cluster, create a role
- In the local cluster, bind a user to that role through a role binding
- Delete the role binding through kubectl
- Delete the role though kubectl
- Verify that the role was deleted
- Verify that the role binding was deleted
- In the local cluster, create a cluster role
- In the local cluster, bind a user to that cluster role though a cluster role binding
- In the local cluster, bind a user to that cluster role through a role binding
- Delete the role binding through kubectl
- Delete the cluster role binding through kubectl
- Delete the cluster role though kubectl
- Verify that the role binding was deleted
- Verify that the cluster role binding was deleted
- Verify that the cluster role was deleted


## What areas could experience regressions

- The ability to remove users from cluster level permissions. Since this code results in new finalizers on our RBAC objects, this could result in rolebindings/clusterrolebindings not being deleted. This could result in more permissions for users than was desired. Note that since this effects the RBAC primitives, not our ClusterRoleTemplateBindings or ProjectRoleTemplateBindings, you will need to verify that the RBAC primitives are deleted. It is not enough to verify that the CRTB or PRTB is deleted. In addition, keep in mind that some permissions are granted to users which have roles in a project in a cluster. 

- The ability to delete RBAC objects in the local cluster. This is related to the first bullet point, but lower level. The way that the solution is structured results in a new finalizer on every Role, Cluster Role, Role Binding, or Cluster Role Binding in the local cluster. The code was configured, essentially, to ignore resources which do not have a `cluster.cattle.io/name` annotation (and a `cluster.cattle.io/namespace` in the case of some RBAC resources). But it's possible that there exists a bug which causes the finalizer to be overly active, so tests should be conducted to ensure basic create/delete of the above resources continues to work without issue in the local cluster. 

## Are the repro steps accurate/minimal?

Yes, they are included here for convenience. 
- Run rancher v2.6.8
- Create a new user with `standard user` global role
- Create a new cluster, as admin
- Add the user to the cluster as a cluster owner
- Open a new window where you can see the existing cluster as the standard user (in Cluster Manager). Place that side by side with the admin user (also in Cluster Manager).
- Initiate a delete of the cluster
- Note that the admin user no longer sees the cluster, but a non-admin user continues to see the cluster (note: this should not happen with the fix)

**Comment by anupama2501** [comment]
Re opening the issue

On v2.6-head a3b5237  and rancher server upgraded from v2.6.8 >> v2.6-head a3b5237 docker install

1. As an admin, create node driver rke1 clusters, k8s `v1.24.4-rancher1-1` , `v1.23.10-rancher1-1` and rke2 clusters, k8s  `v1.23.10+rke2r1`, `v1.24.4+rke2r1`
2. As an admin, create 2 standard users and add the users as cluster owners in both the clusters
3. Login as the cluster owner in a different browser and keep the browser for admin open.
4. Delete the rke1 cluster as an admin and notice the cluster is removed for both admin and cluster owner
5. Delete the rke2 cluster as an admin
7.  Notice the cluster is removed in the admin browser but not on the cluster owner browser with the message on the cluster `Waiting for cluster-api cluster [anurke2-az] to delete` on a fresh install rancher.
8. on the upgraded rancher server the following message is seen `Resource scheduled for deletion`

Some errors in docker logs of the rancher server:
```
2022/09/14 13:00:48 [ERROR] error syncing 'fleet-default/import-token-anurke2az-ad-2b144580-e402-4122-bef5-90985a5-f046c': handler auth-prov-v2-rb: Operation cannot be fulfilled on rolebindings.rbac.authorization.k8s.io "import-token-anurke2az-ad-2b144580-e402-4122-bef5-90985a5-f046c": StorageError: invalid object, Code: 4, Key: /registry/rolebindings/fleet-default/import-token-anurke2az-ad-2b144580-e402-4122-bef5-90985a5-f046c, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: bb9333af-cce2-4a16-bca1-eed56755fb88, UID in object meta: , requeuing
```

**Comment by MbolotSuse** [comment]
Noe: This issue was merged prior to the split between 2.6 and 2.7.

**Comment by anupama2501** [comment]
Validated the following test cases on v2.6-head 28a7473

Pre-reqs:
- As an admin/standard user created both downstream RKE1 and RKE2 node drivers
- Enabled openldap auth

**Test case1:**

- Added the users from openldap as CO,CM,Custom Manage nodes /Manage members in both the clusters as an admin
- Deleted the clusters as admin and opened 4 browser sessions for admin, CO, CM and Custom permissions
- All roles, role bindings, prtb, crtb, cluster roles, cluster role bindings were deleted in the downstream cluster and local cluster

**Test case2:**

- Added the users from openldap as CO, CM,Custom Manage nodes /Manage members in both the clusters as u1 as a standard user 
- Deleted the clusters as u1 and opened 4 browser sessions for admin, CO, CM and Custom permissions
- All roles, role bindings, crtb, cluster roles, cluster role bindings were deleted in the local cluster downstream cluster and local cluster


**Test case3:**
- As an admin create a project - p1
- Added the users from openldap as PO, PM,Custom Manage nodes /Manage members and read-only permissions in both the clusters as admin
- Deleted P1 as admin and opened 4 browser sessions for admin, PO, PM, read-only and Custom permissions
- All roles, role bindings, prtb, crtb, cluster role bindings were deleted from the downstream cluster and local cluster

**Test case4:**
- As a cluster owner u1 with global standard user permissions create a project - p1
- Added the users from openldap as PO, PM,Custom Manage nodes /Manage members and read-only permissions in both the clusters as u1
- Deleted P1 as u1 and opened 4 browser sessions for admin, PO, PM, read-only and Custom permissions
- All roles, role bindings, prtb, crtb, cluster role bindings were deleted from the downstream cluster and local cluster

**Comment by anupama2501** [comment]
Few notes:

1. For both rke1 and rke2 clusters - user with custom permissions have the clusters stayed active until the UI was refreshed
2. On rke2 clusters, cluster owners and cluster member have the cluster stuck in removing while its removed by a cluster Owner.
3. On rke1 clusters cluster members  state of the cluster while deleting wasn’t changed
4. For project permissions - PO, PM ,read-only, custom permissions - cluster status stayed in active state until the UI was refreshed. - On both rke1 and rke2 clusters.

**Comment by anupama2501** [comment]
Verified upgrade cases:

Test case6:

- As an admin/standard user created both downstream RKE1 and RKE2 node drivers
- Created two standard users user1 and user2
- Added the users from openldap as CO,CM in both the clusters as an admin
- Deleted the clusters as admin and opened 3 browser sessions for admin, CO, CM and Custom permissions
- All roles, role bindings, prtb, crtb, cluster roles, cluster role bindings were deleted in the downstream cluster and local cluster

Test case7:

- Added the users as CO, CM in both the clusters as u1 as a standard user
- Deleted the clusters as u1 and opened 3 browser sessions for admin, CO, CM 
- All roles, role bindings, crtb, cluster roles, cluster role bindings were deleted in the local cluster downstream cluster and local cluster

On rke2 clusters, cluster owners and cluster member have the cluster stuck in removing while its removed by an admin

**Comment by anupama2501** [comment]
Since the above test cases work, we are not reverting the fix but bumping the milestone to v2.6.10 for the re opened test case mentioned [here](https://github.com/rancher/rancher/issues/36072#issuecomment-1246747751)

**Comment by gaktive** [comment]
@samjustus the `awaiting-ui` is label is based on waiting on https://github.com/rancher/dashboard/issues/5997, correct? The split tickets for 2.6.x and 2.7.x introduce some confusion here but based on the other ticket, it appears to be the case. Let me know if there's anything else y'all are awaiting from UI.

**Comment by samjustus** [comment]
@gaktive correct, afaik this is the only thing

**Comment by nwmac** [comment]
@samjustus @gaktive https://github.com/rancher/dashboard/issues/5997 has been fixed and the code merged. QA is testing it. This ticket should now be unblocked.

**Comment by samjustus** [comment]
sure-4408

**Comment by samjustus** [comment]
backport won't happen
